benchmarks/src/main/java/io/grpc/benchmarks/qps/QpsAsyncClient.java

```diff
@@ -0,0 +1,411 @@
+    final TestServiceStub stub = TestServiceGrpc.newStub(channel);
+    final Histogram histogram = new Histogram(HISTOGRAM_MAX_VALUE, HISTOGRAM_PRECISION);
+    final HistogramFuture future = new HistogramFuture(histogram);
+
+    stub.unaryCall(request, new StreamObserver<SimpleResponse>() {
+      long lastCall = System.nanoTime();
+
+      @Override
+      public void onValue(SimpleResponse value) {
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        Status status = Status.fromThrowable(t);
+        System.err.println("Encountered an error in unaryCall. Status is " + status);
+        t.printStackTrace();
+
+        future.cancel(true);
+      }
+
+      @Override
+      public void onCompleted() {
+        long now = System.nanoTime();
+        // Record the latencies in microseconds
+        histogram.recordValue((now - lastCall) / 1000);
+        lastCall = now;
+
+        if (endTime > now) {
+          stub.unaryCall(request, this);
+        } else {
+          future.done();
+        }
+      }
+    });
+
+    return future;
+  }
+
+  private static Future<Histogram> doStreamingCalls(Channel channel, final SimpleRequest request,
+                                             final long endTime) {
+    final TestServiceStub stub = TestServiceGrpc.newStub(channel);
+    final Histogram histogram = new Histogram(HISTOGRAM_MAX_VALUE, HISTOGRAM_PRECISION);
+    final HistogramFuture future = new HistogramFuture(histogram);
+
+    ThisIsAHackStreamObserver responseObserver =
+        new ThisIsAHackStreamObserver(request, histogram, future, endTime);
+
+    StreamObserver<SimpleRequest> requestObserver = stub.streamingCall(responseObserver);
+    responseObserver.requestObserver = requestObserver;
+    requestObserver.onValue(request);
+    return future;
+  }
+
+  /**
+   * This seems necessary as we need to reference the requestObserver in the responseObserver.
+   * The alternative would be to use the channel layer directly.
+   */
+  private static class ThisIsAHackStreamObserver implements StreamObserver<SimpleResponse> {
+
+    final SimpleRequest request;
+    final Histogram histogram;
+    final HistogramFuture future;
+    final long endTime;
+    long lastCall = System.nanoTime();
+
+    StreamObserver<SimpleRequest> requestObserver;
+
+    ThisIsAHackStreamObserver(SimpleRequest request,
+                              Histogram histogram,
+                              HistogramFuture future,
+                              long endTime) {
+      this.request = request;
+      this.histogram = histogram;
+      this.future = future;
+      this.endTime = endTime;
+    }
+
+    @Override
+    public void onValue(SimpleResponse value) {
+      long now = System.nanoTime();
+      // Record the latencies in microseconds
+      histogram.recordValue((now - lastCall) / 1000);
+      lastCall = now;
+
+      if (endTime > now) {
+        requestObserver.onValue(request);
```

Comment:
So we're only making one request at a time per channel?  Would we ever want to be able to just send a batch of requests all at once?